[MNG-8418] regression: maven 4: Duplicate key default

**[Jörg Hohwiller](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hohwille)** opened **[MNG-8418](https://issues.apache.org/jira/browse/MNG-8418?redirect=false)** and commented

I am testing migration of a complex project (600+ modules) to maven 4.0.

While I am fully aware that maven 4.0.0-rc1 is not supposed to be production ready, I already want to do further testing to report blockers from real-life projects that could prevent a smooth upgrade from maven 3 (3.9.9) to maven 4 (once 4.0.0 official release is published).

The build (invocation was `{}mvn install -DskipTests -X{`}) is working with maven 3 (tested with 3.8.6, 3.9.6 and 3.9.9) but it failes with maven 4 (also with latest 4.0.0-rc-1):

```java
[ERROR] Failed to execute goal on project myproject.foundation.client: Could not collect dependencies for project com.customer:myproject.foundation.client:jar:4.25.2.0-SNAPSHOT
[ERROR] java.lang.IllegalStateException: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
[ERROR]         Caused by: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
[ERROR] java.lang.IllegalStateException: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
[ERROR]         Caused by: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
[ERROR] : Failed to collect dependencies at com.customer.mylibrary:mylibrary-clientcomponents:jar:3.39.0.0 -> com.miglayout:miglayout:jar:3.7.2
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal on project myproject.foundation.client: Could not collect dependencies for project com.customer:myproject.foundation.client:jar:4.25.2.0-SNAPSHOT
java.lang.IllegalStateException: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
        Caused by: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
java.lang.IllegalStateException: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
        Caused by: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies(LifecycleDependencyResolver.java:279)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectArtifacts(LifecycleDependencyResolver.java:208)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies(LifecycleDependencyResolver.java:128)
    at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved(MojoExecutor.java:368)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:307)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:214)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:179)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:168)
    at org.apache.maven.buildcache.BuildCacheMojosExecutionStrategy.execute(BuildCacheMojosExecutionStrategy.java:145)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:107)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder.lambda$createBuildCallable$1(MultiThreadedBuilder.java:191)
    at java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
    at java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:840)
Caused by: org.apache.maven.project.DependencyResolutionException: Could not collect dependencies for project com.customer:myproject.foundation.client:jar:4.25.2.0-SNAPSHOT
java.lang.IllegalStateException: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
        Caused by: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
java.lang.IllegalStateException: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
        Caused by: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)    at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve(DefaultProjectDependenciesResolver.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies(LifecycleDependencyResolver.java:260)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectArtifacts(LifecycleDependencyResolver.java:208)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies(LifecycleDependencyResolver.java:128)
    at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved(MojoExecutor.java:368)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:307)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:214)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:179)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:168)
    at org.apache.maven.buildcache.BuildCacheMojosExecutionStrategy.execute(BuildCacheMojosExecutionStrategy.java:145)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:107)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder.lambda$createBuildCallable$1(MultiThreadedBuilder.java:191)
    at java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
    at java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:840)
Caused by: org.eclipse.aether.collection.DependencyCollectionException: Failed to collect dependencies at com.customer.mylibrary:mylibrary-clientcomponents:jar:3.39.0.0 -> com.miglayout:miglayout:jar:3.7.2
    at org.eclipse.aether.internal.impl.collect.DependencyCollectorDelegate.collectDependencies(DependencyCollectorDelegate.java:260)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:79)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.collectDependencies(DefaultRepositorySystem.java:241)
    at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve(DefaultProjectDependenciesResolver.java:156)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies(LifecycleDependencyResolver.java:260)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectArtifacts(LifecycleDependencyResolver.java:208)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies(LifecycleDependencyResolver.java:128)
    at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved(MojoExecutor.java:368)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:307)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:214)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:179)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:168)
    at org.apache.maven.buildcache.BuildCacheMojosExecutionStrategy.execute(BuildCacheMojosExecutionStrategy.java:145)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:107)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder.lambda$createBuildCallable$1(MultiThreadedBuilder.java:191)
    at java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
    at java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:840)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
    at java.util.concurrent.FutureTask.report(FutureTask.java:122)
    at java.util.concurrent.FutureTask.get(FutureTask.java:191)
    at org.eclipse.aether.internal.impl.collect.bf.BfDependencyCollector.processDependency(BfDependencyCollector.java:237)
    at org.eclipse.aether.internal.impl.collect.bf.BfDependencyCollector.doCollectDependencies(BfDependencyCollector.java:202)
    at org.eclipse.aether.internal.impl.collect.DependencyCollectorDelegate.collectDependencies(DependencyCollectorDelegate.java:222)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:79)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.collectDependencies(DefaultRepositorySystem.java:241)
    at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve(DefaultProjectDependenciesResolver.java:156)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies(LifecycleDependencyResolver.java:260)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectArtifacts(LifecycleDependencyResolver.java:208)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies(LifecycleDependencyResolver.java:128)
    at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved(MojoExecutor.java:368)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:307)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:214)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:179)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:168)
    at org.apache.maven.buildcache.BuildCacheMojosExecutionStrategy.execute(BuildCacheMojosExecutionStrategy.java:145)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:107)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder.lambda$createBuildCallable$1(MultiThreadedBuilder.java:191)
    at java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
    at java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.IllegalStateException: Duplicate key default (attempted merging values org.apache.maven.api.model.Activation@1556c428 and org.apache.maven.api.model.Activation@62ed65df)
    at java.util.stream.Collectors.duplicateKeyException(Collectors.java:135)
    at java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:182)
    at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
    at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
    at java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(AbstractList.java:720)
    at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
    at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
    at org.apache.maven.internal.impl.model.DefaultModelBuilder.getProfileActivations(DefaultModelBuilder.java:1878)
    at org.apache.maven.internal.impl.model.DefaultModelBuilder$ModelBuilderSessionState.activateFileModel(DefaultModelBuilder.java:1135)
    at org.apache.maven.internal.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readEffectiveModel(DefaultModelBuilder.java:1154)
    at org.apache.maven.internal.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildEffectiveModel(DefaultModelBuilder.java:807)
    at org.apache.maven.internal.impl.model.DefaultModelBuilder$ModelBuilderSessionImpl.build(DefaultModelBuilder.java:231)
    at org.apache.maven.internal.impl.resolver.DefaultArtifactDescriptorReader.loadPom(DefaultArtifactDescriptorReader.java:207)
    at org.apache.maven.internal.impl.resolver.DefaultArtifactDescriptorReader.readArtifactDescriptor(DefaultArtifactDescriptorReader.java:103)
    at org.eclipse.aether.internal.impl.collect.DependencyCollectorDelegate.resolveCachedArtifactDescriptor(DependencyCollectorDelegate.java:484)
    at org.eclipse.aether.internal.impl.collect.bf.BfDependencyCollector.resolveDescriptorForVersion(BfDependencyCollector.java:463)
    at org.eclipse.aether.internal.impl.collect.bf.BfDependencyCollector.lambda$resolveArtifactDescriptorAsync$1(BfDependencyCollector.java:437)
    at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
    at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
    at org.eclipse.aether.internal.impl.collect.bf.BfDependencyCollector.lambda$resolveArtifactDescriptorAsync$4(BfDependencyCollector.java:436)
    at java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:840)
[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn [args] -r
[DEBUG] Shutting down adapter factory; available factories [file-lock, rwlock-local, semaphore-local, noop]; available name mappers [discriminating, file-gav, file-hgav, file-static, gav, static]
[DEBUG] Shutting down 'file-lock' factory
[DEBUG] Shutting down 'rwlock-local' factory
[DEBUG] Shutting down 'semaphore-local' factory
[DEBUG] Shutting down 'noop' factory 
```

I already checked the POM of `{}mylibrary-clientcomponents:jar:3.39.0.0{`}. There are 4 parent POMs involved from where the dependency to `com.miglayout:miglayout:jar:3.7.2` is defined but it is not duplicated anywhere. Since the error indicates activations that IMHO only exist in profiles I also checked all profiles with activations. All profiles share a unique ID. The activations in the mylibrary POMs are always:

```java
<activation>
  <activeByDefault>false</activeByDefault>
</activation> 
```

In the `myproject.foundation.client:jar:4.25.2.0-SNAPSHOT` POM there is also a profile with a different ID that has this activation:

```java
    <profile>
      <id>macos</id>
      <activation>
        <os>
          <family>mac</family>
        </os>
      </activation>
      ... 
```

First thing that comes to my mind would be to implement a helpful implementation of `toString()` for `{}org.apache.maven.api.model.Activation{`}.
NOTE: This issue is not related to concurrency/threading (no `-T` option).

The bug comes from here:

https://github.com/apache/maven/blob/d81b1b6a2870dc954d6d642500ad38e8e4cb3dde/impl/maven-impl/src/main/java/org/apache/maven/internal/impl/model/DefaultModelBuilder.java#L1876-L1878

Since the code already changed on master for rc2 you can immediately close this issue, but I could not find a MNG bug issue for "Duplicate key default".


---

**Affects:** 4.0.0-rc-1

For the record:
* I removed C
* I renamed the IDs in A and B

Now I rerun the build with Maven 4 and still get the same error.
So what is the exception referring to then?
There is no `<id>default</id>` left in any of my POMs.

p.s.: Creating/extracting a reproducer is extremely hard. Especially when I do not understand what the root-cause is... We have over 600 modules so currently I am searching the needle in the haystack.
In general I observed that error messages are so much better in maven 3 compared to maven 4 in order to trace down errors.
Example:
Maven 4:
```
$ mvn install
[INFO] Scanning for projects...
[ERROR] Internal error: org.apache.maven.api.services.MavenException: Multiple fatal exceptions occurred -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: org.apache.maven.api.services.MavenException: Multiple fatal exceptions occurred
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:157)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.doExecute(MavenInvoker.java:452)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:97)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:81)
    at org.apache.maven.cling.invoker.LookupInvoker.doInvoke(LookupInvoker.java:165)
    at org.apache.maven.cling.invoker.LookupInvoker.invoke(LookupInvoker.java:134)
    at org.apache.maven.cling.ClingSupport.run(ClingSupport.java:76)
    at org.apache.maven.cling.MavenCling.main(MavenCling.java:51)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:569)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:314)
Caused by: org.apache.maven.api.services.MavenException: Multiple fatal exceptions occurred
    at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildBuildPom(DefaultModelBuilder.java:805)
    at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionImpl.build(DefaultModelBuilder.java:242)
    at org.apache.maven.project.DefaultProjectBuilder$BuildSession.build(DefaultProjectBuilder.java:544)
    at org.apache.maven.project.DefaultProjectBuilder$BuildSession.lambda$doBuild$6(DefaultProjectBuilder.java:521)
    at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
    at java.util.Collections$2.tryAdvance(Collections.java:4854)
    at java.util.Collections$2.forEachRemaining(Collections.java:4862)
    at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
    at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
    at org.apache.maven.project.DefaultProjectBuilder$BuildSession.doBuild(DefaultProjectBuilder.java:523)
    at org.apache.maven.project.DefaultProjectBuilder$BuildSession.build(DefaultProjectBuilder.java:497)
    at org.apache.maven.project.DefaultProjectBuilder.build(DefaultProjectBuilder.java:194)
    at org.apache.maven.project.collector.DefaultProjectsSelector.selectProjects(DefaultProjectsSelector.java:61)
    at org.apache.maven.project.collector.RequestPomCollectionStrategy.collectProjects(RequestPomCollectionStrategy.java:49)
    at org.apache.maven.graph.DefaultGraphBuilder.getProjectsForMavenReactor(DefaultGraphBuilder.java:367)
    at org.apache.maven.graph.DefaultGraphBuilder.build(DefaultGraphBuilder.java:100)
    at org.apache.maven.DefaultMaven.buildGraph(DefaultMaven.java:643)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:250)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:225)
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:149)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.doExecute(MavenInvoker.java:452)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:97)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:81)
    at org.apache.maven.cling.invoker.LookupInvoker.doInvoke(LookupInvoker.java:165)
    at org.apache.maven.cling.invoker.LookupInvoker.invoke(LookupInvoker.java:134)
    at org.apache.maven.cling.ClingSupport.run(ClingSupport.java:76)
    at org.apache.maven.cling.MavenCling.main(MavenCling.java:51)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:569)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:314)
    Suppressed: java.lang.IllegalArgumentException: Invalid Version Range Request: com.example:project.parent:pom:${revision}${changelist} < [nexus2 (https://repository.maven.org/repository/maven/, default, releases)]
        at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.mayThrow(DefaultRepositorySystemValidator.java:62)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.validateVersionRangeRequest(DefaultRepositorySystemValidator.java:107)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveVersionRange(DefaultRepositorySystem.java:196)
        at org.apache.maven.impl.DefaultVersionRangeResolver.doResolve(DefaultVersionRangeResolver.java:68)
        at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
        at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
        at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
        at org.apache.maven.impl.DefaultVersionRangeResolver.resolve(DefaultVersionRangeResolver.java:60)
        at org.apache.maven.api.services.VersionRangeResolver.resolve(VersionRangeResolver.java:81)
        at org.apache.maven.impl.AbstractSession.resolveHighestVersion(AbstractSession.java:997)
        at org.apache.maven.impl.resolver.DefaultModelResolver.doResolveModel(DefaultModelResolver.java:143)
        at org.apache.maven.impl.resolver.DefaultModelResolver.lambda$resolveModel$0(DefaultModelResolver.java:114)
        at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
        at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
        at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
        at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:114)
        at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:63)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveAndReadParentExternally(DefaultModelBuilder.java:1237)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveParent(DefaultModelBuilder.java:1057)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readParent(DefaultModelBuilder.java:1012)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readEffectiveModel(DefaultModelBuilder.java:1355)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildEffectiveModel(DefaultModelBuilder.java:931)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.lambda$buildBuildPom$6(DefaultModelBuilder.java:782)
        at org.apache.maven.impl.util.PhasingExecutor.lambda$execute$0(PhasingExecutor.java:80)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.lang.Thread.run(Thread.java:840)
        Suppressed: java.lang.IllegalArgumentException: Not fully interpolated artifact com.example:project.parent:pom:${revision}${changelist}
            at org.apache.maven.impl.resolver.validator.MavenValidator.validateArtifact(MavenValidator.java:44)
            at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.validateVersionRangeRequest(DefaultRepositorySystemValidator.java:95)
            at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveVersionRange(DefaultRepositorySystem.java:196)
            at org.apache.maven.impl.DefaultVersionRangeResolver.doResolve(DefaultVersionRangeResolver.java:68)
            at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
            at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
            at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
            at org.apache.maven.impl.DefaultVersionRangeResolver.resolve(DefaultVersionRangeResolver.java:60)
            at org.apache.maven.api.services.VersionRangeResolver.resolve(VersionRangeResolver.java:81)
            at org.apache.maven.impl.AbstractSession.resolveHighestVersion(AbstractSession.java:997)
            at org.apache.maven.impl.resolver.DefaultModelResolver.doResolveModel(DefaultModelResolver.java:143)
            at org.apache.maven.impl.resolver.DefaultModelResolver.lambda$resolveModel$0(DefaultModelResolver.java:114)
            at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
            at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
            at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
            at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:114)
            at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:63)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveAndReadParentExternally(DefaultModelBuilder.java:1237)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveParent(DefaultModelBuilder.java:1057)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readParent(DefaultModelBuilder.java:1012)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readEffectiveModel(DefaultModelBuilder.java:1355)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildEffectiveModel(DefaultModelBuilder.java:931)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.lambda$buildBuildPom$6(DefaultModelBuilder.java:782)
            at org.apache.maven.impl.util.PhasingExecutor.lambda$execute$0(PhasingExecutor.java:80)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
            at java.lang.Thread.run(Thread.java:840)
    Suppressed: java.lang.IllegalArgumentException: Invalid Version Range Request: com.example:project.parent:pom:${revision}${changelist} < [nexus2 (https://repository.maven.org/repository/maven/, default, releases)]
        at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.mayThrow(DefaultRepositorySystemValidator.java:62)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.validateVersionRangeRequest(DefaultRepositorySystemValidator.java:107)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveVersionRange(DefaultRepositorySystem.java:196)
        at org.apache.maven.impl.DefaultVersionRangeResolver.doResolve(DefaultVersionRangeResolver.java:68)
        at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
        at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
        at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
        at org.apache.maven.impl.DefaultVersionRangeResolver.resolve(DefaultVersionRangeResolver.java:60)
        at org.apache.maven.api.services.VersionRangeResolver.resolve(VersionRangeResolver.java:81)
        at org.apache.maven.impl.AbstractSession.resolveHighestVersion(AbstractSession.java:997)
        at org.apache.maven.impl.resolver.DefaultModelResolver.doResolveModel(DefaultModelResolver.java:143)
        at org.apache.maven.impl.resolver.DefaultModelResolver.lambda$resolveModel$0(DefaultModelResolver.java:114)
        at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
        at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
        at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
        at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:114)
        at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:63)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveAndReadParentExternally(DefaultModelBuilder.java:1237)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveParent(DefaultModelBuilder.java:1057)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readParent(DefaultModelBuilder.java:1012)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readEffectiveModel(DefaultModelBuilder.java:1355)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildEffectiveModel(DefaultModelBuilder.java:931)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.lambda$buildBuildPom$6(DefaultModelBuilder.java:782)
        at org.apache.maven.impl.util.PhasingExecutor.lambda$execute$0(PhasingExecutor.java:80)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.lang.Thread.run(Thread.java:840)
        Suppressed: java.lang.IllegalArgumentException: Not fully interpolated artifact com.example:project.parent:pom:${revision}${changelist}
            at org.apache.maven.impl.resolver.validator.MavenValidator.validateArtifact(MavenValidator.java:44)
            at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.validateVersionRangeRequest(DefaultRepositorySystemValidator.java:95)
            at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveVersionRange(DefaultRepositorySystem.java:196)
            at org.apache.maven.impl.DefaultVersionRangeResolver.doResolve(DefaultVersionRangeResolver.java:68)
            at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
            at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
            at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
            at org.apache.maven.impl.DefaultVersionRangeResolver.resolve(DefaultVersionRangeResolver.java:60)
            at org.apache.maven.api.services.VersionRangeResolver.resolve(VersionRangeResolver.java:81)
            at org.apache.maven.impl.AbstractSession.resolveHighestVersion(AbstractSession.java:997)
            at org.apache.maven.impl.resolver.DefaultModelResolver.doResolveModel(DefaultModelResolver.java:143)
            at org.apache.maven.impl.resolver.DefaultModelResolver.lambda$resolveModel$0(DefaultModelResolver.java:114)
            at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
            at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
            at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
            at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:114)
            at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:63)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveAndReadParentExternally(DefaultModelBuilder.java:1237)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveParent(DefaultModelBuilder.java:1057)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readParent(DefaultModelBuilder.java:1012)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readEffectiveModel(DefaultModelBuilder.java:1355)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildEffectiveModel(DefaultModelBuilder.java:931)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.lambda$buildBuildPom$6(DefaultModelBuilder.java:782)
            at org.apache.maven.impl.util.PhasingExecutor.lambda$execute$0(PhasingExecutor.java:80)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
            at java.lang.Thread.run(Thread.java:840)
    Suppressed: java.lang.IllegalArgumentException: Invalid Version Range Request: com.example:project.parent:pom:${revision}${changelist} < [nexus2 (https://repository.maven.org/repository/maven/, default, releases)]
        at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.mayThrow(DefaultRepositorySystemValidator.java:62)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.validateVersionRangeRequest(DefaultRepositorySystemValidator.java:107)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveVersionRange(DefaultRepositorySystem.java:196)
        at org.apache.maven.impl.DefaultVersionRangeResolver.doResolve(DefaultVersionRangeResolver.java:68)
        at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
        at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
        at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
        at org.apache.maven.impl.DefaultVersionRangeResolver.resolve(DefaultVersionRangeResolver.java:60)
        at org.apache.maven.api.services.VersionRangeResolver.resolve(VersionRangeResolver.java:81)
        at org.apache.maven.impl.AbstractSession.resolveHighestVersion(AbstractSession.java:997)
        at org.apache.maven.impl.resolver.DefaultModelResolver.doResolveModel(DefaultModelResolver.java:143)
        at org.apache.maven.impl.resolver.DefaultModelResolver.lambda$resolveModel$0(DefaultModelResolver.java:114)
        at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
        at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
        at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
        at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:114)
        at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:63)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveAndReadParentExternally(DefaultModelBuilder.java:1237)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveParent(DefaultModelBuilder.java:1057)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readParent(DefaultModelBuilder.java:1012)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readEffectiveModel(DefaultModelBuilder.java:1355)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildEffectiveModel(DefaultModelBuilder.java:931)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.lambda$buildBuildPom$6(DefaultModelBuilder.java:782)
        at org.apache.maven.impl.util.PhasingExecutor.lambda$execute$0(PhasingExecutor.java:80)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.lang.Thread.run(Thread.java:840)
        Suppressed: java.lang.IllegalArgumentException: Not fully interpolated artifact com.example:project.parent:pom:${revision}${changelist}
            at org.apache.maven.impl.resolver.validator.MavenValidator.validateArtifact(MavenValidator.java:44)
            at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.validateVersionRangeRequest(DefaultRepositorySystemValidator.java:95)
            at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveVersionRange(DefaultRepositorySystem.java:196)
            at org.apache.maven.impl.DefaultVersionRangeResolver.doResolve(DefaultVersionRangeResolver.java:68)
            at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
            at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
            at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
            at org.apache.maven.impl.DefaultVersionRangeResolver.resolve(DefaultVersionRangeResolver.java:60)
            at org.apache.maven.api.services.VersionRangeResolver.resolve(VersionRangeResolver.java:81)
            at org.apache.maven.impl.AbstractSession.resolveHighestVersion(AbstractSession.java:997)
            at org.apache.maven.impl.resolver.DefaultModelResolver.doResolveModel(DefaultModelResolver.java:143)
            at org.apache.maven.impl.resolver.DefaultModelResolver.lambda$resolveModel$0(DefaultModelResolver.java:114)
            at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
            at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
            at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
            at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:114)
            at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:63)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveAndReadParentExternally(DefaultModelBuilder.java:1237)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveParent(DefaultModelBuilder.java:1057)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readParent(DefaultModelBuilder.java:1012)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readEffectiveModel(DefaultModelBuilder.java:1355)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildEffectiveModel(DefaultModelBuilder.java:931)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.lambda$buildBuildPom$6(DefaultModelBuilder.java:782)
            at org.apache.maven.impl.util.PhasingExecutor.lambda$execute$0(PhasingExecutor.java:80)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
            at java.lang.Thread.run(Thread.java:840)
    Suppressed: java.lang.IllegalArgumentException: Invalid Version Range Request: com.example:project.parent:pom:${revision}${changelist} < [nexus2 (https://repository.maven.org/repository/maven/, default, releases)]
        at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.mayThrow(DefaultRepositorySystemValidator.java:62)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.validateVersionRangeRequest(DefaultRepositorySystemValidator.java:107)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveVersionRange(DefaultRepositorySystem.java:196)
        at org.apache.maven.impl.DefaultVersionRangeResolver.doResolve(DefaultVersionRangeResolver.java:68)
        at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
        at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
        at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
        at org.apache.maven.impl.DefaultVersionRangeResolver.resolve(DefaultVersionRangeResolver.java:60)
        at org.apache.maven.api.services.VersionRangeResolver.resolve(VersionRangeResolver.java:81)
        at org.apache.maven.impl.AbstractSession.resolveHighestVersion(AbstractSession.java:997)
        at org.apache.maven.impl.resolver.DefaultModelResolver.doResolveModel(DefaultModelResolver.java:143)
        at org.apache.maven.impl.resolver.DefaultModelResolver.lambda$resolveModel$0(DefaultModelResolver.java:114)
        at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
        at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
        at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
        at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:114)
        at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:63)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveAndReadParentExternally(DefaultModelBuilder.java:1237)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveParent(DefaultModelBuilder.java:1057)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readParent(DefaultModelBuilder.java:1012)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readEffectiveModel(DefaultModelBuilder.java:1355)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildEffectiveModel(DefaultModelBuilder.java:931)
        at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.lambda$buildBuildPom$6(DefaultModelBuilder.java:782)
        at org.apache.maven.impl.util.PhasingExecutor.lambda$execute$0(PhasingExecutor.java:80)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.lang.Thread.run(Thread.java:840)
        Suppressed: java.lang.IllegalArgumentException: Not fully interpolated artifact com.example:project.parent:pom:${revision}${changelist}
            at org.apache.maven.impl.resolver.validator.MavenValidator.validateArtifact(MavenValidator.java:44)
            at org.eclipse.aether.internal.impl.DefaultRepositorySystemValidator.validateVersionRangeRequest(DefaultRepositorySystemValidator.java:95)
            at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveVersionRange(DefaultRepositorySystem.java:196)
            at org.apache.maven.impl.DefaultVersionRangeResolver.doResolve(DefaultVersionRangeResolver.java:68)
            at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
            at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
            at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
            at org.apache.maven.impl.DefaultVersionRangeResolver.resolve(DefaultVersionRangeResolver.java:60)
            at org.apache.maven.api.services.VersionRangeResolver.resolve(VersionRangeResolver.java:81)
            at org.apache.maven.impl.AbstractSession.resolveHighestVersion(AbstractSession.java:997)
            at org.apache.maven.impl.resolver.DefaultModelResolver.doResolveModel(DefaultModelResolver.java:143)
            at org.apache.maven.impl.resolver.DefaultModelResolver.lambda$resolveModel$0(DefaultModelResolver.java:114)
            at org.apache.maven.impl.cache.CachingSupplier.apply(CachingSupplier.java:50)
            at org.apache.maven.impl.cache.AbstractRequestCache.request(AbstractRequestCache.java:63)
            at org.apache.maven.impl.AbstractSession.request(AbstractSession.java:171)
            at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:114)
            at org.apache.maven.impl.resolver.DefaultModelResolver.resolveModel(DefaultModelResolver.java:63)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveAndReadParentExternally(DefaultModelBuilder.java:1237)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.resolveParent(DefaultModelBuilder.java:1057)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readParent(DefaultModelBuilder.java:1012)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.readEffectiveModel(DefaultModelBuilder.java:1355)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.buildEffectiveModel(DefaultModelBuilder.java:931)
            at org.apache.maven.impl.model.DefaultModelBuilder$ModelBuilderSessionState.lambda$buildBuildPom$6(DefaultModelBuilder.java:782)
            at org.apache.maven.impl.util.PhasingExecutor.lambda$execute$0(PhasingExecutor.java:80)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
            at java.lang.Thread.run(Thread.java:840)
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the '-e' switch
[ERROR] Re-run Maven using the '-X' switch to enable verbose output
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/InternalErrorException
```

So WTH?

With Maven3 it gets obvious what I did wrong:
```
$ mvn install
[INFO] Scanning for projects...
Downloading from nexus2: https://repository.maven.org/repository/maven/com/example/project.parent/$%7Brevision%7D$%7Bchangelist%7D/project.parent-$%7Brevision%7D$%7Bchangelist%7D.pom
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[WARNING] 'parent.relativePath' of POM com.example:project.aggregator:${revision}${changelist} (D:\projects\example\workspaces\main\Maven4-Module-Profile-Bug\pom.xml) points at com.examplex:project.parent instead of com.example:project.parent, please verify your project structure @ line 5, column 11
[FATAL] Non-resolvable parent POM for com.example:project.aggregator:${revision}${changelist}: The following artifacts could not be resolved: com.example:project.parent:pom:${revision}${changelist} (absent): Could not transfer artifact com.example:project.parent:pom:${revision}${changelist} from/to nexus2 (https://repository.maven.org/repository/maven/): status code: 400, reason phrase: Invalid repository path (400) and 'parent.relativePath' points at wrong local POM @ line 5, column 11
 @
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]
[ERROR]   The project com.example:project.aggregator:${revision}${changelist} (D:\projects\example\workspaces\main\Maven4-Module-Profile-Bug\pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for com.example:project.aggregator:${revision}${changelist}: The following artifacts could not be resolved: com.example:project.parent:pom:${revision}${changelist} (absent): Could not transfer artifact com.example:project.parent:pom:${revision}${changelist} from/to nexus2 (https://repository.maven.org/repository/maven/): status code: 400, reason phrase: Invalid repository path (400) and 'parent.relativePath' points at wrong local POM @ line 5, column 11 -> [Help 2]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
```

Ah, now it is clear what I did wrong and how to fix it.
Actually I had a typo in the parents POM `groupId` that I made whilst trying to create a reproducer...

--- a/impl/maven-impl/src/main/java/org/apache/maven/impl/model/DefaultModelBuilder.java
+++ b/impl/maven-impl/src/main/java/org/apache/maven/impl/model/DefaultModelBuilder.java
@@ -1397,7 +1397,7 @@ public class DefaultModelBuilder implements ModelBuilder {
             setSource(inputModel);
             inputModel = modelNormalizer.mergeDuplicates(inputModel, request, this);
 
-            Map<String, Activation> interpolatedActivations = getProfileActivations(inputModel);
+            List<Activation> interpolatedActivations = getProfileActivations(inputModel);
             inputModel = injectProfileActivations(inputModel, interpolatedActivations);
 
             // profile injection
@@ -2367,20 +2367,21 @@ public class DefaultModelBuilder implements ModelBuilder {
                 model);
     }
 
-    private Map<String, Activation> getProfileActivations(Model model) {
+    private List<Activation> getProfileActivations(Model model) {
         return model.getProfiles().stream()
-                .filter(p -> p.getActivation() != null)
-                .collect(Collectors.toMap(Profile::getId, Profile::getActivation));
+                .map(Profile::getActivation)
+                .collect(Collectors.toList());
     }
 
-    private Model injectProfileActivations(Model model, Map<String, Activation> activations) {
+    private Model injectProfileActivations(Model model, List<Activation> activations) {
         List<Profile> profiles = new ArrayList<>();
         boolean modified = false;
-        for (Profile profile : model.getProfiles()) {
+        for (int i = 0; i < model.getProfiles().size(); i++) {
+            Profile profile = model.getProfiles().get(i);
             Activation activation = profile.getActivation();
             if (activation != null) {
                 // restore activation
-                profile = profile.withActivation(activations.get(profile.getId()));
+                profile = profile.withActivation(activations.get(i));
                 modified = true;
             }
             profiles.add(profile);

--- a/impl/maven-impl/src/main/java/org/apache/maven/impl/model/DefaultModelBuilder.java
+++ b/impl/maven-impl/src/main/java/org/apache/maven/impl/model/DefaultModelBuilder.java
@@ -2368,9 +2368,7 @@ public class DefaultModelBuilder implements ModelBuilder {
     }
 
     private List<Activation> getProfileActivations(Model model) {
-        return model.getProfiles().stream()
-                .map(Profile::getActivation)
-                .collect(Collectors.toList());
+        return model.getProfiles().stream().map(Profile::getActivation).collect(Collectors.toList());
     }
 
     private Model injectProfileActivations(Model model, List<Activation> activations) {

--- a/impl/maven-impl/src/test/java/org/apache/maven/impl/model/DefaultModelBuilderTest.java
+++ b/impl/maven-impl/src/test/java/org/apache/maven/impl/model/DefaultModelBuilderTest.java
@@ -29,6 +29,7 @@ import org.apache.maven.api.RemoteRepository;
 import org.apache.maven.api.Session;
 import org.apache.maven.api.model.Dependency;
 import org.apache.maven.api.model.Model;
+import org.apache.maven.api.model.Profile;
 import org.apache.maven.api.model.Repository;
 import org.apache.maven.api.services.ModelBuilder;
 import org.apache.maven.api.services.ModelBuilderRequest;
@@ -150,6 +151,32 @@ class DefaultModelBuilderTest {
         result = builder.newSession().build(request);
         assertNotNull(result);
         assertEquals("0.2.0", result.getEffectiveModel().getVersion());
+    }
+
+    @Test
+    public void testDuplicateProfileIdsRetainActivations() {
+        ModelBuilderRequest request = ModelBuilderRequest.builder()
+                .session(session)
+                .requestType(ModelBuilderRequest.RequestType.CONSUMER_DEPENDENCY)
+                .source(Sources.resolvedSource(
+                        getPom("duplicate-profile-ids"), "org.apache.maven.test:duplicate-profile-ids:1.0.0"))
+                .build();
+        ModelBuilderResult result = assertDoesNotThrow(() -> builder.newSession().build(request));
+        assertNotNull(result);
+
+        List<Profile> profiles = result.getEffectiveModel().getProfiles();
+        assertEquals(2, profiles.size());
+        assertEquals("default", profiles.get(0).getId());
+        assertEquals("default", profiles.get(1).getId());
+        assertNotNull(profiles.get(0).getActivation());
+        assertNotNull(profiles.get(1).getActivation());
+        assertTrue(profiles.get(0).getActivation().isActiveByDefault());
+        assertEquals(
+                "duplicate.profile",
+                profiles.get(1).getActivation().getProperty().getName());
+        assertEquals(
+                "enabled",
+                profiles.get(1).getActivation().getProperty().getValue());
     }
 
     @Test

--- a/impl/maven-impl/src/test/java/org/apache/maven/impl/model/DefaultModelBuilderTest.java
+++ b/impl/maven-impl/src/test/java/org/apache/maven/impl/model/DefaultModelBuilderTest.java
@@ -161,7 +161,8 @@ class DefaultModelBuilderTest {
                 .source(Sources.resolvedSource(
                         getPom("duplicate-profile-ids"), "org.apache.maven.test:duplicate-profile-ids:1.0.0"))
                 .build();
-        ModelBuilderResult result = assertDoesNotThrow(() -> builder.newSession().build(request));
+        ModelBuilderResult result =
+                assertDoesNotThrow(() -> builder.newSession().build(request));
         assertNotNull(result);
 
         List<Profile> profiles = result.getEffectiveModel().getProfiles();
@@ -174,9 +175,7 @@ class DefaultModelBuilderTest {
         assertEquals(
                 "duplicate.profile",
                 profiles.get(1).getActivation().getProperty().getName());
-        assertEquals(
-                "enabled",
-                profiles.get(1).getActivation().getProperty().getValue());
+        assertEquals("enabled", profiles.get(1).getActivation().getProperty().getValue());
     }
 
     @Test

--- a/impl/maven-impl/src/test/resources/poms/factory/duplicate-profile-ids.xml
+++ b/impl/maven-impl/src/test/resources/poms/factory/duplicate-profile-ids.xml
@@ -1,0 +1,47 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!---
+ Licensed to the Apache Software Foundation (ASF) under one or more
+ contributor license agreements.  See the NOTICE file distributed with
+ this work for additional information regarding copyright ownership.
+ The ASF licenses this file to You under the Apache License, Version 2.0
+ (the "License"); you may not use this file except in compliance with
+ the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.1.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.1.0 http://maven.apache.org/xsd/maven-4.1.0.xsd"
+         root="true">
+    <modelVersion>4.1.0</modelVersion>
+
+    <groupId>org.apache.maven.test</groupId>
+    <artifactId>duplicate-profile-ids</artifactId>
+    <version>1.0.0</version>
+    <packaging>pom</packaging>
+
+    <profiles>
+        <profile>
+            <id>default</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
+            </activation>
+        </profile>
+        <profile>
+            <id>default</id>
+            <activation>
+                <property>
+                    <name>duplicate.profile</name>
+                    <value>enabled</value>
+                </property>
+            </activation>
+        </profile>
+    </profiles>
+
+</project>

